torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -254,34 +277,74 @@ def _post_backward(self, *unused: Any):
             )
 
     def finalize_backward(self):
-        if self._sharded_state == ShardedState.UNSHARDED:
-            # Run post-backward here since the forward inputs did not require
-            # gradient, so the post-backward hook did not run
-            self._post_backward()
         if self._reduce_scatter_view_out_event is not None:
             torch.cuda.current_stream().wait_event(self._reduce_scatter_view_out_event)
             self._reduce_scatter_view_out_event = None
         self._training_state = TrainingState.IDLE
+        self._post_forward_indices.clear()
+        self.all_forward_grad_fns.clear()
+
+    def _prefetch_unshard(self):
+        if self._training_state == TrainingState.PRE_BACKWARD:
+            if not self._post_forward_indices:
+                # Can be cleared if running multiple `backward`s
+                return
+            curr_index = self._post_forward_indices.pop()
+            if (target_index := curr_index - 1) < 0:
```

Comment:
This means the root state where it does not have reshard_after_forward, so we don't need the prefetch unshard?

Reply:
Yep! (In general, it is probably good to have the out-of-bounds index check, but yes, in this case, it would be the root.)